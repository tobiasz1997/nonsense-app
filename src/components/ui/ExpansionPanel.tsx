import React, { FC, PropsWithChildren, useState } from 'react';
import { ChevronDownIcon } from '@heroicons/react/24/solid';
import cx from 'classnames';
import Divider from '@components/ui/Divider';
import { AnimatePresence, motion } from 'framer-motion';

type Props = {
	label: string;
};

const ExpansionPanel: FC<PropsWithChildren & Props> = (props) => {
	const [isOpen, setIsOpen] = useState(false);
	return (
		<div>
			<div className="flex items-center">
				<Divider />
				<button
					className="flex items-center gap-x-1 p-2 font-bold text-orange hover:text-yellow"
					onClick={() => setIsOpen((prev) => !prev)}
				>
					{props.label}
					<ChevronDownIcon
						className={cx('h-5 w-5 transition-all', {
							'rotate-180': isOpen
						})}
					/>
				</button>
				<Divider className="w-10" />
			</div>
			<AnimatePresence>
				{isOpen && (
					<motion.div
						initial={{ opacity: 0, height: '0' }}
						animate={{ opacity: 1, height: '100%' }}
						exit={{ opacity: 0, height: '0' }}
					>
						{props.children}
					</motion.div>
				)}
			</AnimatePresence>
		</div>
	);
};

export default ExpansionPanel;
