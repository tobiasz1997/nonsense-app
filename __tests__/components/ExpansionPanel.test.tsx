import { fireEvent, render, screen } from '@testing-library/react';
import ExpansionPanel from '@components/ui/ExpansionPanel';

describe('Expansion Panel', () => {
	const label = 'Test';
	const content = 'content';

	it('should render without crashing', () => {
		const { container } = render(<ExpansionPanel label={label} />);
		expect(container).toMatchSnapshot();
	});

	it('should title be correct', () => {
		render(<ExpansionPanel label={label} />);

		const labelInDoc = screen.getByText(label);

		expect(labelInDoc).toBeInTheDocument();
	});

	it('should content be not visible', () => {
		render(<ExpansionPanel label={label}>{content}</ExpansionPanel>);

		const contentInDoc = screen.queryByText(content);

		expect(contentInDoc).not.toBeInTheDocument();
	});

	it('should content be visible', () => {
		render(<ExpansionPanel label={label}>{content}</ExpansionPanel>);

		fireEvent.click(screen.getByText(label));
		const contentInDoc = screen.queryByText(content);

		expect(contentInDoc).toBeInTheDocument();
	});
});
