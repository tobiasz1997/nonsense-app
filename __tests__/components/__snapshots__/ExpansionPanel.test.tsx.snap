// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expansion Panel should render without crashing 1`] = `
<div>
  <div>
    <div
      class="flex items-center"
    >
      <div
        class="h-0.5 rounded-sm bg-orange w-full"
      />
      <button
        class="flex items-center gap-x-1 rounded p-2 font-bold text-orange focus-visible:outline-none focus-visible:ring-4 focus-visible:ring-orange"
      >
        Test
        <svg
          aria-hidden="true"
          class="h-5 w-5 transition-all"
          fill="currentColor"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            clip-rule="evenodd"
            d="M12.53 16.28a.75.75 0 01-1.06 0l-7.5-7.5a.75.75 0 011.06-1.06L12 14.69l6.97-6.97a.75.75 0 111.06 1.06l-7.5 7.5z"
            fill-rule="evenodd"
          />
        </svg>
      </button>
      <div
        class="h-0.5 rounded-sm bg-orange w-10"
      />
    </div>
  </div>
</div>
`;
